dev-guide/api-conventions.md

```diff
@@ -129,6 +129,125 @@ By providing quality documentation within the API itself, a number of generated
 benefit from the additional context provided which in turn makes it easier for end users to understand
 and use our products.
 
+### Discriminated Unions
+
+In configuration APIs, we commonly have sections of the API model that are only valid to be configured in certain scenarios. A common example of this within OpenShift is platform specific configuration.
+When running on AWS, the other platform configuration blocks are not valid to be set.
+To model this within an API, we use a discriminated union, which models an at-most-one-of semantic with a
+declarative choice.
+
+#### What is a Discriminated Union?
+
+A discriminated union is a structure within the API that closely resembles a union type.
+A number of fields exist within the structure and we are expecting the user to configure precisely one of
+the fields.
+
+In particular, for a discriminated union, an extra field exists which allows the user to declaratively
+state which of particular fields they are configuring.
+
+We use discriminated unions in Kubernetes APIs so that we force the user to make a choice.
+We do not want our code to guess what the user meant, they should tell us which of the choices they made
+using the discriminant.
+
+#### Writing a union in Go
+
+Union types in Go have some additional helper tags which signify how the structure should be handled to consumers.
+Below is an example based on the idea of platform types.
+
+```go
+// MyPlatformConfig is a discriminated union of platform specific configuration.
+// It has a +union tag which informs consumers that this is expected to be a union type.
+// +union
+type MyPlatformConfig struct {
+  // PlatformType is the unions discriminator.
+  // Users are expected to set this value to the name of the platform.
+  // The value of this field should match exactly one field in the union structure.
+  // It has a +unionDiscriminator tag to inform consumers that this is the discriminator field.
+  // The field should be an enum type, so you may also need an enum tag.
+  // The enum values should be in PascalCase.
+  // The field should be required.
+  // +unionDiscriminator
+  // +kubebuilder:validation:Enum:="AWS";"Azure";"GCP"
```

Comment:
We also allow empty string to mean NoOpinion in openshift configuration APIs and allow the platform to choose a good meaning which is subject to change over time.

Reply:
From my understanding this is context specific, for example we have always recommended that a union be on its own (not embedded with other unions) which means that, if you don't care about the union, you can just omit the entire thing. I've certainly seen that in the ControlPlaneMachineSet API as an alternative to having a "NoOpinion". 

I'll add a note to say that you may want to include a no opinion for the discriminant, but I don't think we need it in every case (we can discuss when you're back from PTO)